homeassistant/components/fan/reproduce_state.py

```diff
@@ -33,14 +33,23 @@
 _LOGGER = logging.getLogger(__name__)
 
 VALID_STATES = {STATE_ON, STATE_OFF}
-ATTRIBUTES = {  # attribute: service
+
+SIMPLE_ATTRIBUTES = {  # attribute: service
     ATTR_DIRECTION: SERVICE_SET_DIRECTION,
     ATTR_OSCILLATING: SERVICE_OSCILLATE,
+}
+
+# These need to be handled separately because they are mutually exclusive.
+# The order of these is important, as SPEED is legacy and should be checked for
+# first.
+SPEED_AND_MODES_ATTRIBUTES = {
     ATTR_SPEED: SERVICE_SET_SPEED,
     ATTR_PERCENTAGE: SERVICE_SET_PERCENTAGE,
     ATTR_PRESET_MODE: SERVICE_SET_PRESET_MODE,
 }
 
+ATTRIBUTES = set(SIMPLE_ATTRIBUTES.keys()) | set(SPEED_AND_MODES_ATTRIBUTES)
```

Comment:
My impression is that `set` may not preserve order, which i worry about given the comment above:
```
>>> a = {'a': 2, 'b': 3, 'c': 4}
>>> a.keys()
dict_keys(['a', 'b', 'c'])
>>> set(a.keys())
{'b', 'a', 'c'}
```

Is that a problem?